src/main/java/org/terasology/dynamicCities/buildings/BuildingManager.java

```diff
@@ -188,28 +182,47 @@ public void postBegin() {
 
     public Optional<GenericBuildingComponent> getRandomBuildingOfZoneForCulture(String zone, Rect2i shape, CultureComponent cultureComponent) {
         if (buildings.containsKey(zone)) {
-            int max = buildings.get(zone).size();
-            GenericBuildingComponent building;
+            // TODO: needs to be adapted if buildings have a specific spawn chance
+            List<GenericBuildingComponent> availableBuildings =
+                    buildings.get(zone).stream().filter(b -> cultureComponent.availableBuildings.contains(b.name)).collect(Collectors.toList());
+            int numberOfBuildingsForZone = availableBuildings.size();
+
+            GenericBuildingComponent selectedBuilding = null;
             int iter = 0;
-            do {
-                int index = rng.nextInt(max);
-                building = (GenericBuildingComponent) buildings.get(zone).toArray()[index];
```

Comment:
This created a new array in each iteration. In addition, you can get rid of the type cast by doing something like:
```Java
GenericBuildingComponent[] bldgs = buildings.get(zone).toArray(new GenericBuildingComponent[max])
```